File: Advanced Algorithms/rc4.c
#include <stdio.h>
#include <string.h>

#define N 256

// Swap function
void swap(unsigned char *a, unsigned char *b) {
    unsigned char temp = *a;
    *a = *b;
    *b = temp;
}

// Key Scheduling Algorithm (KSA)
void KSA(unsigned char *key, int keylen, unsigned char *S) {
    int i, j = 0;

    for (i = 0; i < N; i++)
        S[i] = i;

    for (i = 0; i < N; i++) {
        j = (j + S[i] + key[i % keylen]) % N;
        swap(&S[i], &S[j]);
    }
}

// Pseudo-Random Generation Algorithm (PRGA)
void PRGA(unsigned char *S, unsigned char *plaintext, unsigned char *ciphertext, int len) {
    int i = 0, j = 0, k, t;

    for (k = 0; k < len; k++) {
        i = (i + 1) % N;
        j = (j + S[i]) % N;

        swap(&S[i], &S[j]);

        t = (S[i] + S[j]) % N;
        unsigned char keyStream = S[t];

        ciphertext[k] = plaintext[k] ^ keyStream;
    }
}

// RC4 function (same for encryption & decryption)
void RC4(unsigned char *key, int keylen, unsigned char *input, unsigned char *output, int len) {
    unsigned char S[N];

    KSA(key, keylen, S);
    PRGA(S, input, output, len);
}

int main() {
    unsigned char key[] = "secretkey";
    unsigned char plaintext[] = "Hello RC4 Cipher";

    int len = strlen((char *)plaintext);

    unsigned char ciphertext[256];
    unsigned char decrypted[256];

    // Encryption
    RC4(key, strlen((char *)key), plaintext, ciphertext, len);

    printf("Ciphertext (in hex): ");
    for (int i = 0; i < len; i++) {
        printf("%02X ", ciphertext[i]);
    }
    printf("\n");

    // Decryption (same RC4 function)
    RC4(key, strlen((char *)key), ciphertext, decrypted, len);
    decrypted[len] = '\0';

    printf("Decrypted text: %s\n", decrypted);

    return 0;
}

File: rc4.py
# RC4 implementation in Python

def KSA(key):
    S = list(range(256))
    j = 0
    key_length = len(key)

    for i in range(256):
        j = (j + S[i] + key[i % key_length]) % 256
        S[i], S[j] = S[j], S[i]

    return S


def PRGA(S, data_len):
    i = 0
    j = 0

    for _ in range(data_len):
        i = (i + 1) % 256
        j = (j + S[i]) % 256

        S[i], S[j] = S[j], S[i]

        K = S[(S[i] + S[j]) % 256]
        yield K


def rc4(key, data):
    key = [ord(c) for c in key]
    S = KSA(key)
    keystream = PRGA(S, len(data))

    return bytes([c ^ k for c, k in zip(data, keystream)])


# ------------------ MAIN ------------------

if __name__ == "__main__":
    key = "secretkey"
    plaintext = "Hello RC4 Cipher"

    # Convert plaintext to bytes
    plaintext_bytes = plaintext.encode()

    # Encryption
    ciphertext = rc4(key, plaintext_bytes)
    print("Ciphertext (hex):", ciphertext.hex())

    # Decryption (same function)
    decrypted = rc4(key, ciphertext)
    print("Decrypted text:", decrypted.decode())